ledger/streaming/client/src/block_validator.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Stream that validates block components
+
+use futures::{future, Stream, StreamExt};
+use mc_common::{
+    logger::{log, Logger},
+    HashSet,
+};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::{
+    BlockStream, BlockStreamComponents, Error as StreamError, Result as StreamResult,
+};
+use mc_transaction_core::{compute_block_id, ring_signature::KeyImage, Block};
+
+/// Create stream factory for validating individual blocks within a stream.
+/// Valid blocks will passed on, blocks that don't pass will pass an error.
+pub struct BlockValidator<US: BlockStream + 'static, L: Ledger + 'static> {
+    upstream: US,
+    start_block: Block,
+    ledger: Option<L>,
+    logger: Logger,
+}
+
+impl<US: BlockStream + 'static, L: Ledger> BlockValidator<US, L> {
+    /// Create new block validation stream
+    pub fn new(upstream: US, start_block: Block, ledger: Option<L>, logger: Logger) -> Self {
+        Self {
+            upstream,
+            start_block,
+            ledger,
+            logger,
+        }
+    }
+}
+
+impl<US: BlockStream + 'static, L: Ledger + Clone + 'static> BlockStream for BlockValidator<US, L> {
+    type Stream = impl Stream<Item = StreamResult<BlockStreamComponents>>;
+
+    /// Get block stream that performs validation
+    fn get_block_stream(&self, starting_height: u64) -> StreamResult<Self::Stream> {
+        //let logger = self.logger.clone();
+        let prev_block_id = self.start_block.id.clone();
+        let ledger = self.ledger.clone();
+        let additional_key_images: HashSet<KeyImage> = HashSet::default();
+
+        log::info!(self.logger, "Creating block validation stream");
+        let stream = self.upstream.get_block_stream(starting_height)?;
+
+        Ok(stream.scan(
+            (ledger, prev_block_id, additional_key_images),
```

Comment:
Consider using a dedicated struct for this. I expect it would make the state tracking easier to follow

Reply:
Agreed